### ISSUE TITLE
Poor numerics in CPU tf.multinomial

### ISSUE THREAD

**Issue by nikitakit** [tensorflow/tensorflow#2774]
The CPU implementation of tf.multinomial seems to have numerical underflow/overflow when logits are outside of a certain narrow range. The usable logit range on my system is about -88 to 88, which is unacceptably small for my use case (where the logits come from an automatically-trained hidden layer).

For example:
```
>>> logits = np.array([[1000.]*5])
>>> sess.run(tf.multinomial(logits, 10))
array([[5, 5, 5, 5, 5, 5, 5, 5, 5, 5]])
```

Note that the index 5 is out of range (which is the _undocumented_ behavior used to signal an error condition). Expected behavior is to sample uniformly from [0, 4], such as in this workaround:

```
>>> logits = np.array([[1000.]*5])
>>> sess.run(tf.multinomial(tf.nn.log_softmax(logits), 10))
array([[2, 0, 4, 1, 0, 0, 2, 2, 2, 4]])
```

It's a similar story for negative logits:
```
>>> logits = np.array([[-1000.]*5])
>>> sess.run(tf.multinomial(logits, 10))
array([[5, 5, 5, 5, 5, 5, 5, 5, 5, 5]])
>>> sess.run(tf.multinomial(tf.nn.log_softmax(logits), 10))
array([[2, 4, 3, 3, 3, 2, 2, 2, 3, 0]])
```
I think tf.multinomial should be fixed to accept a wider range of inputs. Or, if that's not possible e.g. due to speed tradeoffs, the workaround above should be noted in the documentation.

**System info**: tensorflow 845fb7ab77fae53b849894925c7d13c1b8918675, Python 3, CPU only on Mac OSX.

**Comment by girving** [comment]
@concretevitamin Looks like `tf.multinomial` doesn't subtract off the maximum logit before taking exponentials.  Can you fix it to do that?  The undocumented error case is also bad; not sure where in the code it's doing that.

**Comment by concretevitamin** [comment]
I'm looking into this.

**Comment by concretevitamin** [comment]
I pushed a fix via commit [de6ce1de08](https://github.com/tensorflow/tensorflow/commit/de6ce1de08ea97d599687fbbe5196ca4af5232ae).  Please feel free to re-open if related issues come up.  Thanks!

**Comment by girving** [comment]
@concretevitamin In future, please automatically close issues from the commit by saying "Fixes #2774" or similar: https://help.github.com/articles/closing-issues-via-commit-messages

**Comment by concretevitamin** [comment]
Yep, if you look at my commit message, I included something similar but
slightly more complicated for Github to grasp ("fixes github issue
#????").  Will make it simpler next time :)

**Comment by seansaito** [comment]
This still occurs for Python 2, Ubuntu 14.04

**Comment by girving** [comment]
@seansaito We'll need more information to be able to help, including the TensorFlow version.

**Comment by fidlej** [comment]
The out of range index can be still produced, if the input is not finite.
Is it desired? Should it be documented?

Example:
```
>>> sess.run(tf.multinomial([[1e1000] * 5], 10))
array([[5, 5, 5, 5, 5, 5, 5, 5, 5, 5]])
```

**Comment by TobiasLee** [comment]
Yeah, I met this problem, too.

**Comment by nikola-j** [comment]
Mee too.
Simple way to reproduce:
```python
import tensorflow as tf
import numpy as np
isess = tf.InteractiveSession()
dist = tf.multinomial([[np.inf, np.inf]], 1)
isess.run(dist) # outputs 2, instead of 0 or 1
```
Ubuntu 16.04, tensorflow 1.9, python3.6, it also works in python2.7
